### PR TITLE
`New Feature` Termed Roles

### LINKED ISSUE #103: `[Termed Roles]` Create `DecentAutonomousAdminHat` contract

### LINKED ISSUE #104: `[Termed Roles]` Create Tests for `DecentAutonomousAdminHat` contract

### LINKED ISSUE #105: `[Termed Roles]` Update `DecentHat_0_1_0` contract to deploy DecentAAHat as wearer of AdminHat

### PR DESCRIPTION
**Pull request by dacebt**
Adds DecentAutonomousAdmin Contract
- Used for trigging next start and flush any sablier funds to previous elected wearer

Adds DecentHats_0_2_0 Contract
- DecentAutoAdmin clone is deployed and added as wearer of Admin Hat
- Updated to create Termed Roles when prompted
- This involves making the deploying a instance of Hats Election Module and setting the eligibility module

Closes #103 
Closes #104 
Closes #105

### CODE REVIEW

File: contracts/DecentHats_0_1_0.sol
Comment (adamgall) [inline contracts/DecentHats_0_1_0.sol:1]: We're going to need to create another version of this contract now (0_2_0), because 

1. this existing version has already been deployed on networks _and_ has been used to create real roles on real daos
2. this contract's address is part of the salt used to determine role's smartAccount addresses.

So this means we can't just update this and deploy a new one, because then we'll be determining the incorrect smartAccount address for any existing Roles on DAOs that used this current version to create their Roles.

**Review by adamgall — CHANGES_REQUESTED** [review]

File: contracts/DecentHats_0_1_0.sol
Comment (dacebt) [reply contracts/DecentHats_0_1_0.sol:1]: thats a good point. going to have to figure out a good way to enable versioning of our contracts. Noting for later: What does this transaction to 'update' look like.

File: contracts/DecentHats_0_1_0.sol [hunk 1]
```diff
@@ -247,4 +247,4 @@ contract DecentHats_0_1_0 {
 
         params.hatsProtocol.transferHat(topHatId, address(this), msg.sender);
     }
-}
+}
```
Comment (adamgall) [inline contracts/DecentHats_0_1_0.sol]: Please make sure this change is reverted.

File: contracts/interfaces/hats/IHats.sol [hunk 2]
```diff
@@ -39,4 +39,15 @@ interface IHats {
...
+    function getHatEligibilityModule(
+        uint256 _hatId
+    ) external view returns (address eligibility);
+
+    function isWearerOfHat(
+        address _user,
+        uint256 _hatId
+    ) external view returns (bool isWearer);
+
+    function changeHatEligibility(uint256 _hatId, address _newEligibility) external;
```
Comment (adamgall) [inline contracts/interfaces/hats/IHats.sol]: Please see the following commits and discussions from other PRs:

- Relevant discussion, but for Sablier not Hats: https://github.com/decentdao/decent-contracts/pull/100#discussion_r1792349087
- Relevant commit which implements what I think should happen in this branch, too: https://github.com/decentdao/decent-contracts/pull/108/commits/e1fd7d75a9ba03aa9f47c4fef03a8cf3f359eace

Please let me know if it's not clear why we need to do this. The TL;DR is that if we _don't_, then you'll notice that running the deployment scripts will result in a new instance of `DecentHats_0_1_0` being deployed (because the bytecode has changed), and we need to avoid that.

**Review by adamgall — CHANGES_REQUESTED** [review]

File: contracts/interfaces/hats/IHats.sol (hunk 2 above)
Comment (dacebt) [reply contracts/interfaces/hats/IHats.sol]: AH you know I watched for that and got fooled by the caching and hadn't looked at the deployment diff yet ha. cool beans I'll take a look at what you've done and update accordingly

Comment (adamgall) [reply contracts/interfaces/hats/IHats.sol]: I actually implemented it on @DarksightKellar's branch... let me cherry pick a couple of commits over

Comment (dacebt) [reply contracts/interfaces/hats/IHats.sol]: even better 🙏

Comment (adamgall) [reply contracts/interfaces/hats/IHats.sol]: @Da-Colon holy shit it ended up being 17 commits LMAO. Had to revert the changes to both the Hats interfaces _and_ the Sablier interfaces. Luckily, the two other open branches implement the "fixes" for both Hats and Sablier, so more of less that part was already done just had to cherry pick them over.

Then, a lot of other commits happened due to cleanup. But that's only because I made changes very atomically. Recommend you read through them in order to see what I did.

File: contracts/DecentHats_0_2_0.sol [hunk 3]
```diff
@@ -11,7 +11,6 @@ import {LockupLinear, Broker} from "./interfaces/sablier/full/types/DataTypes.so
 import {DecentAutonomousAdmin} from "./DecentAutonomousAdmin.sol";
 import {IHatsModuleFactory} from "./interfaces/hats/full/IHatsModuleFactory.sol";
 import {IHatsElectionEligibility} from "./interfaces/hats/full/IHatsElectionEligibility.sol";
-import {Clones} from "@openzeppelin/contracts/proxy/Clones.sol";
```
Comment (dacebt) [inline contracts/DecentHats_0_2_0.sol]: ty

File: contracts/DecentAutonomousAdmin.sol [hunk 4]
```diff
@@ -1,5 +1,6 @@
 // SPDX-License-Identifier: MIT
-pragma solidity =0.8.19;
+pragma solidity 0.8.28;
```
Comment (dacebt) [inline contracts/DecentAutonomousAdmin.sol:2]: 👍

File: contracts/DecentHats_0_1_0.sol [hunk 5]
```diff
@@ -247,4 +247,4 @@ contract DecentHats_0_1_0 {
 
         params.hatsProtocol.transferHat(topHatId, address(this), msg.sender);
     }
-}
\ No newline at end of file
+}
```
Comment (dacebt) [inline contracts/DecentHats_0_1_0.sol]: 👍

**Review by dacebt — COMMENTED** [review]
coolio changes makes sense. Question though. Should we just change the 'deploy' script to prevent any new deployments of that contract.

Another question. Should we probably think about a different pattern for updates for V2 and beyond?

File: contracts/DecentAutonomousAdmin.sol [hunk 6]
```diff
@@ -0,0 +1,130 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.28;
+
+import {IHats} from "./interfaces/hats/full/IHats.sol";
+import {IHatsElectionEligibility} from "./interfaces/hats/full/IHatsElectionEligibility.sol";
+import {ISablierV2Lockup} from "./interfaces/sablier/full/ISablierV2Lockup.sol";
+
+contract DecentAutonomousAdmin {
+    string public constant NAME = "DecentAutonomousAdmin";
+    string public version_ = "0.1.0";
```
Comment (adamgall) [inline contracts/DecentAutonomousAdmin.sol]: Should probably make this `constant`

Comment (adamgall) [inline contracts/DecentAutonomousAdmin.sol]: But actually...

For this contract, which is a "long standing" contract (aka not an ephemeral contract in the context of a proposal execution that we add/remove as a module to execute specific logic), I'd like to see ERC165 support built in. Will be good for having a programmatic way to determine contract functionality when this contract will inevitably be upgraded in the future, and we may have DAOs with different versions of this contract installed on their Admin role.

I think the ERC165 support should replace the (custom, non-idiomatic) `NAME` and `version_` variables.

File: contracts/DecentHats.sol [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+import {IERC6551Registry} from "./interfaces/IERC6551Registry.sol";
+import {IHats} from "./interfaces/hats/full/IHats.sol";
+import {LockupLinear, Broker} from "./interfaces/sablier/full/types/DataTypes.sol";
+import {DecentAutonomousAdmin} from "./DecentAutonomousAdmin.sol";
+import {IHatsModuleFactory} from "./interfaces/hats/full/IHatsModuleFactory.sol";
+import {IHatsElectionEligibility} from "./interfaces/hats/full/IHatsElectionEligibility.sol";
+import {ModuleProxyFactory} from "@gnosis.pm/zodiac/contracts/factory/ModuleProxyFactory.sol";
+
+contract DecentHats {
+    string public constant NAME = "DecentHats";
```
Comment (adamgall) [inline contracts/DecentHats.sol]: Same comments here. Instead of this NAME variable, I'd like to see ERC165 support built into this contract.

File: deploy/core/011_deploy_ModuleProxyFactory.ts
Comment (adamgall) [inline deploy/core/011_deploy_ModuleProxyFactory.ts:1]: I would prefer you undo the changes to this file.

File: deploy/core/017_deploy_DecentHats_0_1_0.ts
Comment (adamgall) [inline deploy/core/017_deploy_DecentHats_0_1_0.ts:1]: I'd prefer you follow the same pattern as script 011 (after addressing my comment)

File: test/DecentHats_0_1_0.test.ts
Comment (adamgall) [inline test/DecentHats_0_1_0.test.ts:1]: Undo this change please.

**Review by adamgall — CHANGES_REQUESTED** [review]

File: contracts/DecentHats.sol (hunk 7 above)
Comment (adamgall) [reply contracts/DecentHats.sol]: Wait nevermind we don't need ERC165 in here, this is one of those "ephemeral" contracts.

**Review by DarksightKellar — COMMENTED** [review]
Man, I started reviewing this some time ago and I remember I had some questions at some point, but now I'm back to it and I don't remember what. Since this is significant work, I'd love to have an all hands engineering review at some point so we can all, both 
 - come to the same page on what the feature+code is and should look like, and
 - contribute and pick apart in real time, where thoughts will bounce of each other and flow a lot more freely.

File: contracts/DecentAutonomousAdmin.sol (hunk 6 above)
Comment (dacebt) [reply contracts/DecentAutonomousAdmin.sol]: Just saw these comments over the weekend. Let me get this done

File: contracts/DecentAutonomousAdmin.sol [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        address currentWearer;
+        IHats userHatProtocol;
+        uint256 userHatId;
+        address nominatedWearer;
+    }
+
+    // //////////////////////////////////////////////////////////////
+    //                         initializer
+    // //////////////////////////////////////////////////////////////
+    function setUp(bytes memory initializeParams) public override initializer {}
```
Comment (dacebt) [inline contracts/DecentAutonomousAdmin.sol:18]: Uses the same patterns as the modules deployed via `ModuleProxyFactory` which requires a setUp function with params, but in its current state we don't need it

File: contracts/DecentAutonomousAdmin.sol [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+    // //////////////////////////////////////////////////////////////
+    //                         Public Functions
+    // //////////////////////////////////////////////////////////////
+    function triggerStartNextTerm(TriggerStartArgs calldata args) public {
```
Comment (dacebt) [inline contracts/DecentAutonomousAdmin.sol:23]: Pretty straight forward. The Contract allows the Admin Hat, when worn to help trigger the next term. While the next term can be triggered by anyone. we need a few other things to happen when this occurs for UX purposes.

- After starting next term. The Original Hat should be burned via `checkHatWearerStatus`. if ineligibile the hat is burned. Then we just mint a new hat to the new wearer

File: contracts/DecentHats.sol [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+        (uint256 adminHatId, ) = _createAdminHatAndAccount(
+            params.hatsProtocol,
+            params.registry,
+            params.moduleProxyFactory,
+            params.decentAutonomousAdminMasterCopy,
+            params.hatsAccountImplementation,
+            topHatAccount,
+            topHatId,
+            params.adminHat
+        );
```
Comment (dacebt) [inline contracts/DecentHats.sol]: This function was created from `_createHatAndAccountAndMintAndStreams` and narrowed down to just Admin hat creation.

File: contracts/DecentHats.sol [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+                    _createElectionEligiblityModule(
+                        params.hatsModuleFactory,
+                        params.hatsElectionEligibilityImplementation,
+                        params.hatsProtocol.getNextId(adminHatId),
+                        topHatId,
+                        params.hats[i].termedParams[0]
+                    ),
+                    adminHatId,
+                    params.hats[i]
+                );
```
Comment (dacebt) [inline contracts/DecentHats.sol]: This is the other 'new' workflow to review

File: contracts/interfaces/IDecentAutonomousAdmin.sol [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+pragma solidity 0.8.28;
+
+import {IHats} from "./hats/full/IHats.sol";
+import {ISablierV2Lockup} from "./sablier/full/ISablierV2Lockup.sol";
+
+interface IDecentAutonomousAdmin {
+    struct SablierStreamInfo {
+        uint256 streamId;
+        ISablierV2Lockup sablierV2Lockup;
+    }
```
Comment (adamgall) [inline contracts/interfaces/IDecentAutonomousAdmin.sol]: I think this struct (and the `ISablierV2Lockup`) are unused and can be removed

File: contracts/mocks/MockHatsAdmin.sol
Comment (adamgall) [inline contracts/mocks/MockHatsAdmin.sol:1]: Nitpicky but can you rename this file to MockHatsAutoAdmin to match the contract name within

Comment (adamgall) [reply contracts/mocks/MockHatsAdmin.sol:1]: Also, what's the reason for this specific mock contract? It seems like it's supposed to mock the core Hats contract. Any reason why you didn't just add this code to the existing MockHats contract?

**Review by adamgall — CHANGES_REQUESTED** [review]
High-level first-draft review

File: contracts/DecentHats.sol [hunk 13]
```diff
@@ -0,0 +1,393 @@
...
+        for (uint256 i = 0; i < params.hats.length; ) {
+            if (params.hats[i].isTermed) {
+                // Create election module and set as eligiblity
+                _createTermedHatAndAccountAndMintAndStreams(
+                    params.hatsProtocol,
+                    topHatAccount,
+                    _createElectionEligiblityModule(
+                        params.hatsModuleFactory,
+                        params.hatsElectionEligibilityImplementation,
+                        params.hatsProtocol.getNextId(adminHatId),
```
Comment (adamgall) [inline contracts/DecentHats.sol]: Has this been tested when there are multiple termed hats being created?

edit: actually I do think this would work as written, yeah. `getNextId(adminHatId)` is being called after each new hat is created, yeah.

File: test/DecentHats.test.ts
Comment (adamgall) [inline test/DecentHats.test.ts:1]: Nowhere in this test are there any hats being created where `isTermed: true`. That seems like an oversight...

**Review by adamgall — CHANGES_REQUESTED** [review]
Please write some tests for the new DecentHats contract where termed hats are being created.

File: contracts/DecentHats.sol (hunk 13 above)
Comment (adamgall) [reply contracts/DecentHats.sol]: but still, definitely need some tests written

File: contracts/interfaces/IDecentAutonomousAdmin.sol (hunk 12 above)
Comment (dacebt) [reply contracts/interfaces/IDecentAutonomousAdmin.sol]: 🤦 oof yeah good call

File: test/DecentHats.test.ts
Comment (adamgall) [reply test/DecentHats.test.ts:1]: @Da-Colon I remember you saying yesterday that there was a reason why there are no tests testing `isTermed: true`. Can you explain that again?

File: contracts/DecentHats.sol
Comment (adamgall) [inline contracts/DecentHats.sol:1]: Make sure that the code from most recent `DecentHats_0_1_0` gets implemented into this contract.

https://github.com/decentdao/decent-contracts/blob/develop/contracts/DecentHats_0_1_0.sol#L198-L233

@DarksightKellar please provide @Da-Colon with the proper natspec comments, too

**Review by adamgall — CHANGES_REQUESTED** [review]

File: contracts/DecentHats.sol
Comment (dacebt) [reply contracts/DecentHats.sol:1]: done + updated

File: contracts/DecentHats.sol (hunk 13 above)
Comment (dacebt) [reply contracts/DecentHats.sol]: got the tree test. Need to add a test for `createTermedRoleHat` external function

File: contracts/mocks/MockHatsAdmin.sol
Comment (dacebt) [reply contracts/mocks/MockHatsAdmin.sol:1]: I'll take another look at the AutoAdmin test and Mock contracts it uses now with everything updated

File: test/DecentHats.test.ts [hunk 14]
```diff
@@ -0,0 +1,826 @@
+import { SignerWithAddress } from '@nomicfoundation/hardhat-ethers/signers';
+import { expect } from 'chai';
+/* eslint-disable-next-line import/no-extraneous-dependencies */
+import { ethers } from 'ethers';
```
Comment (dacebt) [inline test/DecentHats.test.ts]: @adamgall @DarksightKellar So that 'can't find ethers namespace' errors was removing type safety. After quite a bit of trial and error of trying things. I found this is the simplest solution. `hardhat` or one of the `hardhat` libraries is adding `ethers` to our `node_modules` but it is not in our `package.json` but adding it there conflicts with the version used in the hard hat environment. 

So doing this allows for it to properly find it and doesn't prevent the other errors from showing up. 

I did not add this to other files from my own as I didn't want to fill up this PR but I can.

Comment (adamgall) [reply test/DecentHats.test.ts]: I dig it, thanks. I'll open up a standalone PR.

Comment (adamgall) [reply test/DecentHats.test.ts]: Well actually there are only two other test files that have this issue. Would you mind just applying this fix to those two other files in your current PR? thanks.

Comment (dacebt) [reply test/DecentHats.test.ts]: Ah good call thought it was in more places, but looks like its only effecting where we are trying to pull `types` out of the package

File: test/DecentAutonomousAdmin.test.ts [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+    it('should correctly invalidate random address as current wearer', async function () {
+      const args = {
+        currentWearer: randomUser.address,
+        userHatProtocol: await hatsProtocol.getAddress(),
+        userHatId: userHatId,
+        nominatedWearer: nominatedWearer.address,
+        sablierStreamInfo: [], // No Sablier stream info for this test
+      };
+
+      // Verify the hat is now worn by the current wearer
```
Comment (DarksightKellar) [inline test/DecentAutonomousAdmin.test.ts]: Incorrect copy-paste comment

File: test/DecentAutonomousAdmin.test.ts [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+        userHatId: userHatId,
+        nominatedWearer: nominatedWearer.address,
+        sablierStreamInfo: [], // No Sablier stream info for this test
+      };
+
+      // Call triggerStartNextTerm on the adminHat contract
+      await adminHat.triggerStartNextTerm(args);
+
+      // Verify the hat is now worn by the nominated wearer
+      expect((await hatsProtocol.isWearerOfHat(nominatedWearer.address, userHatId)) === true);
```
Comment (DarksightKellar) [inline test/DecentAutonomousAdmin.test.ts:88]: Suggest adding another test to verify the previous wearer of the hat was "removed"

**Review by DarksightKellar — CHANGES_REQUESTED** [review]

**Review by DarksightKellar — APPROVED** [review]

File: contracts/DecentHats_0_1_0.sol
Comment (adamgall) [inline contracts/DecentHats_0_1_0.sol:1]: Why are there changes in this contract @Da-Colon @DarksightKellar ??

This contract has already been deployed and it's not going to be deployed again (see the commented out deployment code in deployer file 017).

We should not be making changes to contract source code if there's no intention of deploying it. We should make sure that whatever source code is checked into source control always contains the most-recently-deployed contract on public chains.

**Review by adamgall — CHANGES_REQUESTED** [review]

File: contracts/interfaces/IDecentAutonomousAdmin.sol [hunk 17]
```diff
@@ -0,0 +1,16 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.28;
+
+import {IHats} from "./hats/full/IHats.sol";
+import {ISablierV2Lockup} from "./sablier/full/ISablierV2Lockup.sol";
```
Comment (adamgall) [inline contracts/interfaces/IDecentAutonomousAdmin.sol]: Does this import need to be here? It's not being referenced in this interface file, so I would say this is probably not the right place for it.

File: test/DecentHats.test.ts [hunk 18]
```diff
@@ -0,0 +1,970 @@
...
+        expect(hatsCountBeforeCreate).to.equal(2); // Top hat + admin hat
+
+        await createRoleHatPromise;
+
+        const newHatId = await mockHats.count();
+        expect(newHatId).to.equal(3); // + newly created hat
+      });
+    });
+
+    describe('Creating a new hat on existing Tree', () => {
```
Comment (DarksightKellar) [inline test/DecentHats.test.ts]: Should be "Creating a new Termed hat on existing tree"

File: contracts/mocks/MockHats.sol
Comment (adamgall) [inline contracts/mocks/MockHats.sol:1]: Considering that this Mock contract implements the not-"`full`" `IHats` interface, which is specifically _only_ used in an old version of `DecentHats_0_1_0`, I'm wondering why there needs to be additional functionality added to this Mock contract?

File: contracts/DecentHats_0_1_0.sol
Comment (DarksightKellar) [reply contracts/DecentHats_0_1_0.sol:1]: Reverted. The natspec update went onto the wrong contract

File: contracts/mocks/MockHats.sol
Comment (adamgall) [inline contracts/mocks/MockHats.sol:1]: Why do we need additional functionality in this contract? This specific Mock implements the non-"`full`" `IHats` interface, which is only used in the old `DecentHats_0_1_0` contract. I do see that there are failing tests if I remove these new mocked functions, but that leads me to believe that the test file is relying on this `MockHats` instance when maybe it should be relying on the new `MockHatsAutoAdmin` instance.

File: contracts/mocks/MockHatsAdmin.sol
Comment (adamgall) [inline contracts/mocks/MockHatsAdmin.sol:1]: Please make sure the filename of this contract matches the actual contract name.

Comment (adamgall) [inline contracts/mocks/MockHatsAdmin.sol:1]: Is there any reason why this Mock file / contract has the word "Admin" in it?

Maybe a more appropriate name would be something like "MockHatsFull", to refer to the fact that it implements the "full" IHats interface.

**Review by adamgall — CHANGES_REQUESTED** [review]

File: contracts/DecentHats_0_1_0.sol
Comment (adamgall) [reply contracts/DecentHats_0_1_0.sol:1]: Thank you

File: contracts/mocks/MockHatsAdmin.sol
Comment (adamgall) [reply contracts/mocks/MockHatsAdmin.sol:1]: fixed

Comment (adamgall) [reply contracts/mocks/MockHatsAdmin.sol:1]: fixed

File: contracts/mocks/MockHats.sol
Comment (adamgall) [reply contracts/mocks/MockHats.sol:1]: fixed

File: contracts/interfaces/IDecentAutonomousAdmin.sol (hunk 17 above)
Comment (adamgall) [reply contracts/interfaces/IDecentAutonomousAdmin.sol]: fixed

File: test/DecentHats.test.ts (hunk 18 above)
Comment (adamgall) [reply test/DecentHats.test.ts]: fixed

File: contracts/mocks/MockHatsAdmin.sol
Comment (adamgall) [reply contracts/mocks/MockHatsAdmin.sol:1]: fixed

File: contracts/DecentHats.sol (hunk 13 above)
Comment (adamgall) [reply contracts/DecentHats.sol]: addressed

File: contracts/DecentAutonomousAdmin.sol [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+        require(
+            args.hatsProtocol.isWearerOfHat(args.currentWearer, args.hatId),
+            "Not current wearer"
+        );
```
Comment (adamgall) [inline contracts/DecentAutonomousAdmin.sol]: To stick with modern solidity, can we use actual `error`s here?

File: contracts/DecentAutonomousAdmin.sol [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+        address hatsEligibilityModuleAddress = args
+            .hatsProtocol
+            .getHatEligibilityModule(args.hatId);
```
Comment (adamgall) [inline contracts/DecentAutonomousAdmin.sol]: To save some gas, let's not allocate memory for this variable, since it's only being used in one place. Just inline this into line 33.

File: contracts/DecentHats.sol [hunk 21]
```diff
@@ -0,0 +1,485 @@
...
+pragma solidity 0.8.28;
+
+import {Enum} from "@gnosis.pm/safe-contracts/contracts/common/Enum.sol";
+import {IAvatar} from "@gnosis.pm/zodiac/contracts/interfaces/IAvatar.sol";
+import {Strings} from "@openzeppelin/contracts/utils/Strings.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {IERC6551Registry} from "./interfaces/IERC6551Registry.sol";
+import {IHats} from "./interfaces/hats/full/IHats.sol";
+import {LockupLinear, Broker} from "./interfaces/sablier/full/types/DataTypes.sol";
+import {DecentAutonomousAdmin} from "./DecentAutonomousAdmin.sol";
```
Comment (adamgall) [inline contracts/DecentHats.sol]: This import seems unneeded.

File: contracts/DecentHats.sol (hunk 7 above)
Comment (adamgall) [reply contracts/DecentHats.sol]: We also probably just don't even need this NAME string in the contract either. Like who cares, yanno? Idk why I added it in DecentHats_0_1_0.

File: contracts/DecentHats.sol [hunk 22]
```diff
@@ -0,0 +1,485 @@
...
+
+    struct Hat {
+        address wearer;
+        string details;
+        string imageURI;
+        SablierStreamParams[] sablierParams;
+        TermedParams[] termedParams;
+        uint32 maxSupply;
+        bool isMutable;
+        bool isTermed;
```
Comment (adamgall) [inline contracts/DecentHats.sol]: Do we need to have this boolean? Can we not infer it based on the existence of any items in the `termedParams` array?

File: contracts/DecentHats.sol [hunk 23]
```diff
@@ -0,0 +1,485 @@
...
+        uint128 termEndDateTs;
+        address[] nominatedWearers;
+    }
+
+    struct Hat {
+        address wearer;
+        string details;
+        string imageURI;
+        SablierStreamParams[] sablierParams;
+        TermedParams[] termedParams;
```
Comment (adamgall) [inline contracts/DecentHats.sol]: Why is this an array?

File: contracts/DecentHats.sol [hunk 24]
```diff
@@ -0,0 +1,485 @@
...
+        IHatsModuleFactory hatsModuleFactory,
+        address hatsElectionEligibilityImplementation,
+        uint256 hatId,
+        uint256 topHatId,
+        TermedParams calldata termedParams
+    ) internal returns (address electionModuleAddress) {
+        electionModuleAddress = hatsModuleFactory.createHatsModule(
+            hatsElectionEligibilityImplementation,
+            hatId,
+            abi.encode(topHatId, uint256(0)),
```
Comment (adamgall) [inline contracts/DecentHats.sol]: Does this 0 need to be cast? Also, can you please add some comments on these lines about what these "immutable" and "mutable" args represent, from the PoV of the Elections Eligibility contract?

File: contracts/DecentHats.sol [hunk 25]
```diff
@@ -0,0 +1,485 @@
...
+        Hat calldata hat
+    ) internal {
+        require(
+            hat.termedParams.length == 1,
+            "DecentHats: termedParams length must be 1"
+        );
+        require(
+            hat.termedParams[0].nominatedWearers.length == 1,
+            "DecentHats: nominatedWearers length must be 1"
+        );
```
Comment (adamgall) [inline contracts/DecentHats.sol]: Can we use actual `error`s here? And everywhere in this contract.

File: contracts/DecentHats.sol [hunk 26]
```diff
@@ -0,0 +1,485 @@
...
+            topHatAccount,
+            eligibilityAddress
+        );
+
+        IHatsElectionEligibility(eligibilityAddress).elect(
+            hat.termedParams[0].termEndDateTs,
+            hat.termedParams[0].nominatedWearers
+        );
+
+        hatsProtocol.mintHat(hatId, hat.termedParams[0].nominatedWearers[0]);
```
Comment (adamgall) [inline contracts/DecentHats.sol]: Definitely open to pushback on this suggestion, but, what if we called `batchMintHat` here and passed in the entire `nominatedWearers` array? Leave it to the frontend to just make sure we're only passing in one of them. That might allow us to continue using this same contract if we do ever decide to open up "role members" to > 1 at a time.

File: contracts/DecentHats.sol [hunk 27]
```diff
@@ -0,0 +1,485 @@
...
+            adminHatId,
+            hat,
+            topHatAccount,
+            eligibilityAddress
+        );
+
+        IHatsElectionEligibility(eligibilityAddress).elect(
+            hat.termedParams[0].termEndDateTs,
+            hat.termedParams[0].nominatedWearers
+        );
```
Comment (adamgall) [inline contracts/DecentHats.sol]: Nitpick, flip the order of these two statements.

File: contracts/DecentHats.sol (hunk 26 above)
Comment (adamgall) [reply contracts/DecentHats.sol]: Edit: nevermind, this would cause a discrepancy in how we're setting up streams then, cause we are hardcoding the [0] there. And it's beyond us to decide how streams should happen if there are multiple nominated wearers.

So now maybe I'm thinking, can we just turn `nominatedWearers` into a single `address`? We'd need to put that address into an array when calling eligibility.elect, but that's fine. I kind of like that better, making the contract's input params match our business logic.

File: contracts/DecentHats.sol [hunk 28]
```diff
@@ -0,0 +1,481 @@
...
+            _hatsProtocol.createHat(
+                adminHatId,
+                _hat.details,
+                _hat.maxSupply,
+                eligibility,
+                toggle,
+                _hat.isMutable,
+                _hat.imageURI
+            );
+    }
```
Comment (adamgall) [inline contracts/DecentHats.sol]: Is it benefitting us to have this single external function call wrapped in an internal function? Why not just inline the `_hatsProtocol.crateHat()` call wherever it's used?

File: contracts/DecentHats.sol [hunk 29]
```diff
@@ -0,0 +1,481 @@
...
+    ) internal returns (address) {
+        return
+            _registry.createAccount(
+                _hatsAccountImplementation,
+                SALT,
+                block.chainid,
+                protocolAddress,
+                hatId
+            );
+    }
```
Comment (adamgall) [inline contracts/DecentHats.sol]: Same as previous comment

File: contracts/DecentHats.sol [hunk 30]
```diff
@@ -0,0 +1,481 @@
...
+        TermedParam calldata termedParam
+    ) internal returns (address electionModuleAddress) {
+        electionModuleAddress = hatsModuleFactory.createHatsModule(
+            hatsElectionEligibilityImplementation,
+            hatId,
+            abi.encode(topHatId, uint256(0)),
+            abi.encode(termedParam.termEndDateTs),
+            uint256(SALT)
+        );
+    }
```
Comment (adamgall) [inline contracts/DecentHats.sol]: same as previous comment

**Review by adamgall — CHANGES_REQUESTED** [review]

File: contracts/DecentHats.sol (hunk 22 above)
Comment (adamgall) [reply contracts/DecentHats.sol]: fixed, thanks

File: contracts/DecentHats.sol (hunk 7 above)
Comment (adamgall) [reply contracts/DecentHats.sol]: fixed

File: contracts/DecentHats.sol (hunk 21 above)
Comment (adamgall) [reply contracts/DecentHats.sol]: fixed

File: contracts/DecentHats.sol (hunk 23 above)
Comment (adamgall) [reply contracts/DecentHats.sol]: addressed

File: contracts/DecentHats.sol (hunk 26 above)
Comment (adamgall) [reply contracts/DecentHats.sol]: addressed this

File: contracts/DecentAutonomousAdmin.sol (hunk 19 above)
Comment (dacebt) [reply contracts/DecentAutonomousAdmin.sol]: fixed

File: contracts/DecentAutonomousAdmin.sol (hunk 20 above)
Comment (dacebt) [reply contracts/DecentAutonomousAdmin.sol]: fixed

File: contracts/DecentHats.sol (hunk 27 above)
Comment (dacebt) [reply contracts/DecentHats.sol]: ?

Comment (dacebt) [reply contracts/DecentHats.sol]: I guess it doesn't really matter, but it seem to make sense to create the hat then elect?

File: contracts/DecentHats.sol (hunk 28 above)
Comment (dacebt) [reply contracts/DecentHats.sol]: fixed.

File: contracts/DecentHats.sol (hunk 30 above)
Comment (dacebt) [reply contracts/DecentHats.sol]: fixed.

File: contracts/DecentHats.sol (hunk 29 above)
Comment (dacebt) [reply contracts/DecentHats.sol]: fixed.

File: contracts/DecentHats.sol (hunk 27 above)
Comment (adamgall) [reply contracts/DecentHats.sol]: yeahh... my nitpick is just because the `hatId` variable isn't used in the `elect` call, but it's used in the line below that. My intention with the nitpick comment was just to keep relevant code close to each other.

> I guess it doesn't really matter, but it seem to make sense to create the hat then elect?

but the `elect`ion doesn't depend on the new `hatId`, is what i'm saying.

File: contracts/DecentHats.sol [hunk 31]
```diff
@@ -0,0 +1,436 @@
...
+            params.adminHatId,
+            params.hat
+        );
+
+        params.hatsProtocol.transferHat(
+            params.topHatId,
+            address(this),
+            msg.sender
+        );
+    }
```
Comment (adamgall) [inline contracts/DecentHats.sol]: I think the last thing to do in this contract, is clean up this standalone function.

@DarksightKellar this one is on you.

So, the reason why we need to transfer the top back back to the safe, is because _this_ contract is the current wearer of the top hat by the time this function is being called. We did that (transaction was created in the frontend and part of the bundle of txs in the proposal that this call is in) because this contract can't create a new hat on a tree unless it's an admin of the hat it's trying to create. Makes sense.

BUT, instead of transferring the top hat to this contract, why not utilize the fact that we've added this contract as a module on the Safe! As of right now, adding this contract as a module on the Safe (in the context of minting a new hat via this call), _is completely unnecessary_, _unless we actually call `execTransactionFromModule` in this function call_.

So let's do that instead of transferring the top hat around.

Please proxy the call to create a new hat through the Safe via `execTransactionFromModule`, and remove the `transferHat` instruction. In the frontend, no need to transfer the top hat to this contract.

File: contracts/DecentHats.sol [hunk 32]
```diff
@@ -0,0 +1,436 @@
...
```
Comment (adamgall) [inline contracts/DecentHats.sol]: Oh look this function exists too!

@Da-Colon @DarksightKellar see my previous comment in this PR. Same applies here.

File: contracts/DecentHats.sol (hunk 25 above)
Comment (adamgall) [reply contracts/DecentHats.sol]: fixed

**Review by adamgall — CHANGES_REQUESTED** [review]
Last set of change requests, related to the two external standalone functions which create new Hats on existing trees.

Those functions _shouldn't_ need to transfer the top hat back to the Safe, because this contract _shouldn't_ need to take ownership of the top hat in the first place.

Just use the fact that we add this contract as a module on the Safe, to proxy those calls to create new hats through the Safe.

File: contracts/DecentHats.sol (hunk 32 above)
Comment (dacebt) [reply contracts/DecentHats.sol]: Let's talk about this today.

Comment (adamgall) [reply contracts/DecentHats.sol]: ready when you are 🫡

**Review by adamgall — APPROVED** [review]
